drivers/include/mtd_flashpage.h

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (C) 2018 OTA keys S.A.
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_mtd_flashpage   Flashpage MTD
+ * @ingroup     drivers_storage
+ * @brief       Driver for internal flash devices implementing flashpage interface
+ *
+ * @{
+ *
+ * @file
+ * @brief       Interface definition for the flashpage memory driver
+ *
+ * @author      Vincent Dupont <vincent@otakeys.com>
+ */
+
+#ifndef MTD_FLASHPAGE_H
+#define MTD_FLASHPAGE_H
+
+#include "mtd.h"
+#include "periph/flashpage.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#define MTD_FLASHPAGE_INIT_VAL(_pages_per_sector) { \
+    .driver = &mtd_flashpage_driver, \
+    .sector_count = FLASHPAGE_NUMOF, \
+    .pages_per_sector = _pages_per_sector,\
+    .page_size = FLASHPAGE_SIZE / _pages_per_sector,\
+}
```

Comment:
I'm not really sure if this is a good idea... What's the purpose of having this wrapper? The biggest problem I see is that you'd hardly know where you can write without precise knowledge about the non-filled flash space left when flashing a firmware... The main purpose of the flashpage driver is to write new program data, or extend the current one (new elf files, scripts of any kind, etc.), and, if I understand correctly, this driver aims to write user data...

Reply:
The goal is to be able to use a file system on internal flash.

It's then completely up to the developer to be sure of what's he's doing and it might be dangerous, of course.

Reply:
I see the point, but I have the issue of erasing. How do you cope with the small erases case? You need to always backup the whole block in RAM before adding something, then erase the whole block and write it from the buffer...

Reply:
I don't understand your point here. This is the case with any flash, either internal or not.
A file system designed to run on a flash will not try to erase a few bytes, but always the entire sector. What is different here?

Reply:
Most SPI flashes have a sector size of 4KB, or even 64KB. And a typical "page" size (i.e. maximum number of bytes you can write with a single instruction) is 256B on these flashes.

Then yes, we rely on the file system to manage erases and write cycles.

Reply:
So, in overall (a bit out of scope now) using ANY file system over a flash-based MTD device will take *at least* as much RAM as the smallest erasable block is it? I just checked and for the N25Q128 series (the one attached to the iotlab-m3) has 4KB as eraseable sub-sector.

Reply:
If you take the assumption that a sector will be modified, this will be true. But a good filesystem won't need to modify a byte in the middle of a sector. See littles or spiffs, they don't need a buffer of the size of a sector.